### PR TITLE
feat(headers): canonical header ordering + get_header_id()

### PR DESCRIPTION
**Pull request by LFRezende**
**Implementation Details:**

Following the PR-SPLIT PLAN (https://github.com/HathorNetwork/hathor-core/blob/feat/shielded-outputs/docs/plans/shielded-pr-split.md), this is STEP 3.

**Alterations:**

Adds an abstract get_header_id() -> bytes classmethod on VertexBaseHeader, and concrete implementations on FeeHeader (0x11) and NanoHeader (0x10) in both hathor-core and hathorlib. Enforces canonical  (ascending-by-ID) header ordering at verification time so a transaction with multiple headers has exactly one valid byte representation, closing a malleability gap.

Defensive add in dag_builder/vertex_exporter.py sorts headers by ID after building, so future contributors adding new add_*_if_needed calls can't accidentally introduce a non-canonical order.

Chain-data scan confirmed no historical mainnet/testnet tx with multiple headers violates the ascending-order invariant, so the rule ships unconditionally (no feature flag, no Feature Activation gate).

**Tests:**
  - hathor_tests/tx/test_canonical_header_order.py: positive +
  negative cases for FeeHeader/NanoHeader ordering, single-header and zero-header edge cases, and direct get_header_id() byte-value assertions.
  - hathor_tests/tx/test_nano_header.py: FakeHeader test fixture now
    implements get_header_id() (returning the sentinel 0xff) so the
    abstract-method ABC machinery doesn't break existing tests.
  - hathorlib/tests/test_basic.py::test_header_ids: hathorlib mirror.

  No external wire-format change; only header order is constrained.

**PR for the scan confirmation:**

https://github.com/HathorNetwork/hathor-core/pull/1695

### CODE REVIEW

File: hathor_tests/tx/test_canonical_header_order.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        vertex = MagicMock()
+        vertex.headers = [nano, fee]
+        vertex.get_maximum_number_of_headers = MagicMock(return_value=3)
+
+        params = MagicMock()
+
+        with patch.object(VertexVerifier, 'get_allowed_headers', return_value={NanoHeader, FeeHeader}):
+            # Should not raise
+            verifier.verify_headers(vertex, params)
+
```
Comment (LFRezende) [inline hathor_tests/tx/test_canonical_header_order.py]: Not sure I'm comfortable with tests being done by raiseErrors in inner functions

File: hathor/dag_builder/vertex_exporter.py [hunk 2]
```diff
@@ -336,6 +336,12 @@ def add_headers_if_needed(self, node: DAGNode, vertex: BaseTransaction) -> None:
         """Add the configured headers."""
         self.add_nano_header_if_needed(node, vertex)
         self.add_fee_header_if_needed(node, vertex)
+        # Future-proofing against header-ordering footguns: sort the headers by
+        # their canonical ID byte so any new add_*_if_needed call added later
+        # cannot accidentally produce a non-canonical order that would then be
+        # rejected by VertexVerifier.verify_headers. The current add order
+        # (nano 0x10, fee 0x11) is already canonical, so this is a no-op today.
+        vertex.headers.sort(key=lambda h: h.get_header_id())
```
Comment (LFRezende) [inline hathor/dag_builder/vertex_exporter.py]: Add comment: `# Ensure headers are in canonical (ascending VertexHeaderId) order.`

File: hathor_tests/tx/test_canonical_header_order.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
+# Copyright 2026 Hathor Labs
```
Comment (LFRezende) [inline hathor_tests/tx/test_canonical_header_order.py]: Why build a new file for the canonical header order? We can do so in the test_shielded_cons_fiixes.py file. I guess we could modularize it further, though. Is this filee created later in the pr-plan and we'll move it there thence?

File: hathor_tests/tx/test_canonical_header_order.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        settings = MagicMock(spec=HathorSettings)
+        verifier = VertexVerifier(settings=settings, reactor=MagicMock(), feature_service=MagicMock())
+
+        # Wrong order: FeeHeader (0x11) before NanoHeader (0x10).
+        nano = NanoHeader.__new__(NanoHeader)
+        fee = FeeHeader.__new__(FeeHeader)
+
+        vertex = MagicMock()
+        vertex.headers = [fee, nano]
+        vertex.get_maximum_number_of_headers = MagicMock(return_value=3)
```
Comment (LFRezende) [inline hathor_tests/tx/test_canonical_header_order.py]: Test file not including shielded outputs header. It makes sense, since we'll add it solely on PR-4, but we'll need to move this back to  the original file, and make sure this test receives the ShieldedOutputs Header back.

File: hathor_tests/tx/test_canonical_header_order.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        verifier = VertexVerifier(settings=settings, reactor=MagicMock(), feature_service=MagicMock())
+
+        # NanoHeader is 0x10, FeeHeader is 0x11 — ascending.
+        nano = NanoHeader.__new__(NanoHeader)
+        fee = FeeHeader.__new__(FeeHeader)
+
+        vertex = MagicMock()
+        vertex.headers = [nano, fee]
+        vertex.get_maximum_number_of_headers = MagicMock(return_value=3)
+
```
Comment (LFRezende) [inline hathor_tests/tx/test_canonical_header_order.py]: Doesn't seem to be a problem, but shouldn't here be return_value=2?

File: hathor_tests/tx/test_canonical_header_order.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    def test_non_canonical_order_rejected(self) -> None:
+        """Headers NOT in ascending order should be rejected."""
+        settings = MagicMock(spec=HathorSettings)
+        verifier = VertexVerifier(settings=settings, reactor=MagicMock(), feature_service=MagicMock())
+
+        # Wrong order: FeeHeader (0x11) before NanoHeader (0x10).
+        nano = NanoHeader.__new__(NanoHeader)
+        fee = FeeHeader.__new__(FeeHeader)
+
```
Comment (LFRezende) [inline hathor_tests/tx/test_canonical_header_order.py]: Instead of ShieldedOutputsHeader, for now.

File: hathor_tests/tx/test_canonical_header_order.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        vertex = MagicMock()
+        vertex.headers = []
+        vertex.get_maximum_number_of_headers = MagicMock(return_value=3)
+
+        params = MagicMock()
+
+        with patch.object(VertexVerifier, 'get_allowed_headers', return_value=set()):
+            # Should not raise
+            verifier.verify_headers(vertex, params)
+
```
Comment (LFRezende) [inline hathor_tests/tx/test_canonical_header_order.py]: Extra test. Doesn't hurt.

File: hathor_tests/tx/test_canonical_header_order.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    def test_header_id_classmethod(self) -> None:
+        """Concrete subclasses expose the right 1-byte header ID."""
+        assert FeeHeader.get_header_id() == b'\x11'
+        assert NanoHeader.get_header_id() == b'\x10'
```
Comment (LFRezende) [inline hathor_tests/tx/test_canonical_header_order.py]: This one too I guess.

File: hathorlib/hathorlib/headers/base.py [hunk 9]
```diff
@@ -22,6 +22,12 @@
 
 
 class VertexBaseHeader(ABC):
+    @classmethod
+    @abstractmethod
+    def get_header_id(cls) -> bytes:
+        """Return the 1-byte header ID for this header type."""
+        raise NotImplementedError
+
```
Comment (LFRezende) [inline hathorlib/hathorlib/headers/base.py]: Why implement the get_header_id here? In the VertexBaseHeader, if it is not implemented in the original pr plan?

I get that we may need to adapt later.

File: hathorlib/hathorlib/headers/fee_header.py [hunk 10]
```diff
@@ -40,6 +40,10 @@ class FeeEntry:
 
 @dataclass(frozen=True)
 class FeeHeader(VertexBaseHeader):
+    @classmethod
+    def get_header_id(cls) -> bytes:
+        return VertexHeaderId.FEE_HEADER.value
+
```
Comment (LFRezende) [inline hathorlib/hathorlib/headers/fee_header.py]: Mirroring methods from hathor-core to hathorlib.

I guess this is fine here.

File: hathorlib/tests/test_basic.py [hunk 11]
```diff
@@ -276,3 +276,20 @@ def test_tx_version_and_signal_bits(self):
 
         self.assertEqual(block.signal_bits, block2.signal_bits)
         self.assertEqual(block.version, block2.version)
+
```
Comment (LFRezende) [inline hathorlib/tests/test_basic.py]: Extra test on header_ids. Wasn't fetched, but can be left here.

**Review by LFRezende — COMMENTED** [review]
Review done mostly on structure, no Shielded Transactions angle for me to review.

**Comment by LFRezende** [comment]
Closing this PR.

After internal discussion the team decided not to enforce canonical header ordering at the present moment. The motivation behind this PR (preventing same-logical-tx-multiple-byte-representations via a verifier rule + supporting get_header_id() machinery) was reweighed: the determinism benefit alone doesn't justify the ongoing cost of adding get_header_id() to every existing and future header class, plus the new consensus rejection path it introduces.

The revert commit pushed to this branch (74d7c556) removes all the canonical-order changes (the verifier check, the _get_header_order helper, the abstract method on VertexBaseHeader, concrete implementations on FeeHeader and NanoHeader in both hathor-core and hathorlib, the defensive DAG-builder sort, FakeHeader.get_header_id() in the test fixture, and the dedicated canonical-order test file), restoring the branch to a net-zero diff against master.
